### PR TITLE
[m685] Make BA life_histories a m2m field

### PR DESCRIPTION
**Pull request by saanobhaai**

### CODE REVIEW

**Comment by saanobhaai** [comment]
[make life_history M2M](https://trello.com/c/fumdgb58/685-make-lifehistory-m2m)

File: src/api/admin/mermaid.py [hunk 1]
```diff
@@ -549,28 +555,34 @@ def fk_link(self, obj):
...
     def child_life_histories(self, obj):
         proportions = {}
-        # get all non-null life histories of descendants
-        life_histories = [ba.life_history for ba in obj.descendants if ba.life_history]
-        total = float(len(life_histories))
-        for lh in life_histories:
-            if lh.name not in proportions:
-                proportions[lh.name] = 0
-            proportions[lh.name] += 1 / total
+        counts = {}
```
Comment (gridcell) [inline src/api/admin/mermaid.py:568]: Could use `counts = defaultdict(int)` and then get rid `if lh.name not in counts:`

Comment (saanobhaai) [reply src/api/admin/mermaid.py]: refactored

File: src/api/admin/mermaid.py [hunk 2]
```diff
@@ -549,28 +555,34 @@ def fk_link(self, obj):
...
+        total = 0
+        life_histories_sets = [ba.life_histories.all() for ba in obj.descendants]
+        for lhs in life_histories_sets:
+            for lh in lhs:
+                if lh.name not in counts:
+                    counts[lh.name] = 0
+                counts[lh.name] += 1
+                total += 1
+        for name in counts:
+            proportions[name] = counts[name] / float(total)
```
[the remark this answers is not in the document]
Comment (saanobhaai) [reply src/api/admin/mermaid.py]: refactored

**Review by gridcell — APPROVED** [review]

**Pull request merged**